### DISCUSSION TITLE
On-demand type generation not working when using Docker

### DISCUSSION THREAD

**Discussion by eriknygren** [ElMassimo/types_from_serializers#13] [Q&A]
[troubleshooting section]: https://types-from-serializers.netlify.app/faqs/

- [x] I have tried upgrading by running `bundle update types_from_serializers`.
- [x] I have read the __[troubleshooting section]__ before opening an issue.

### Description 📖

On demand generation on change not working, but I found a workaround that does the job for now.

### Reproduction/Logs 🐞📜

Hi Máximo, thanks for another awesome gem.

I'm struggling a bit with the autogeneration of types. It seems to be initializing properly, and indeed tracking changes too.

If I make a change to a serializer, and then check the output of running `Rails.application.reloaders` in my console, I get the following output:
```ruby
<TypesFromSerializers::Changes:0x0000ffff9862ec68 @added=#<Set: {}>, @modified=#<Set: {"/app/app/typed_serializers/my_serializer.rb"}>, @removed=#<Set: {}>>
```
So the tracking does seem to be working, but it nothing is ever re-generated, unless I explicitly call `TypesFromSerializers.generate_changed` in the rails console, then it works.

So I then tried forcing it by adding another hook in my initializer:
```ruby
  Rails.application.reloader.to_prepare do
    puts "force serializer generation"
    TypesFromSerializers.generate_changed
  end
```
But at this stage, it doesn't seem to be aware of file changes. Like this hook has a different context to my rails console. So nothing happens.

As a workaround for now I changed this hook to:
```ruby
  Rails.application.reloader.to_prepare do
    TypesFromSerializers.generate(force: true)
  end
```
And it works, and it's fast enough re-generating everything that it's an acceptable workaround.

So that's my issue, and I wanted to ask you if you have an idea why this may be happening? Just to mention I have also installed the `listen` gem.

Cheers!

**Comment by ElMassimo** [comment]
Hi Erik! Thanks for reporting.

It would be interesting to know if in your app the [following condition](https://github.com/ElMassimo/types_from_serializers/blob/main/types_from_serializers/lib/types_from_serializers/railtie.rb#L9) is met when the gem is loaded.

You can try [opening the gem](https://maximomussini.com/posts/debugging-ruby-libraries#opening-gems-%E2%9B%8F) by running `bundle open types_from_serializers`, and adding a `binding.pry` or `debug` statement to check if `Rails.env` is available.

The most likely scenario is that this condition is `false`, so maybe the `reloader` and the `to_prepare` hook are not registered in the running server (even though they are registered in the console).

---

> So I then tried forcing it by adding another hook in my initializer

That makes sense, because it's the `TypesFromSerializers.track_changes` reloader that gathers the changes for `generate_changes` to handle it.

**Comment by eriknygren** [comment]
Hi Máximo, thanks for the quick response!

Following your steps, I can see that the `Rails.env` is indeed correctly set, but the problem seems to be that code in the following block does not run:

```ruby
initializer "types_from_serializers.reloader" do |app|
```

I'm not sure if this is due to something being odd with my app's set up that leads to an incorrect load order, and this hook getting registered too late. But after reading [this thread](https://discuss.rubyonrails.org/t/railtie-initializer-executed-on-startup-but-block-is-never-invoked/81365/7), I tried changing the hook mentioned above to:
```ruby
config.after_initialize do |app|
```
and then everything works as expected.

I don't know enough about the inner workings of Rails to say if this is a issue with this gem or the problem is in my app.

**Comment by ElMassimo** [comment]
That thread suggests that by the time the railtie is loaded, `Rails.application.initialize!` might have already been called.

You could try debugging to see in which order the code loads in your app.

Another thing that might be affecting how things load is whether you have this gem in a non-standard group in your Gemfile or something that causes this line not to load the gem:

```ruby
Bundler.require(*Rails.groups)
```

**Comment by eriknygren** [comment]
Hi, apologies the load order might have been a false flag, using docker I struggled a bit to patch the installed version of the gem inside the container but I have now succeeded with that. Putting debug prints in there I can see the initializer firing as expected.

I have now put some console outputs around the generator class, and I can see that  [changes.updated?](https://github.com/ElMassimo/types_from_serializers/blob/main/types_from_serializers/lib/types_from_serializers/generator.rb#L287C22-L287C22) is always false. So it gets to this stage, but never regenerates anything because of that.

So the problem seems to be the changes aren’t being successfully tracked, and I’m not sure exactly why. This is where it gets tricky because if look in the rails console, I can see entries in the `@modified` list if I inspect the rails reloaders, highly confusing!

Possibly the bug could be related to the fact that I've changed the default folder to:
```ruby
    config.serializers_dirs = [Rails.root.join("app/typed_serializers").to_s]
```
but this has not caused any other issues.

If there’s nothing else obvious to try, nothing to worry about as it works pretty well with this hook set up as a workaround
```  ruby
Rails.application.reloader.to_prepare do
  TypesFromSerializers.generate(force: true)
end
```

**Comment by ElMassimo** [comment]
> using docker

It could be related to the volume that is mounted, and whether `listen` is able to detect changes and notify them.

Thanks for sharing a workaround, it should be helpful for other folks using Docker in development.
